flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -4941,6 +5106,27 @@ class ArrayExprLowering {
                                                   slice);
       };
     }
+    if (isBoxValueEle()) {
+      return [=](IterSpace iters) -> ExtValue {
+        auto eleTy = adjustedArraySubtype(arrTy, iters.iterVec());
+        if (eleTy.isa<fir::BoxType>() || eleTy.isa<fir::SequenceType>())
+          eleTy = fir::ReferenceType::get(eleTy);
+        mlir::Value box;
+        if (iters.empty()) {
+          box = mlir::cast<fir::ArrayLoadOp>(arrLd.getDefiningOp()).memref();
+        } else {
+          box = builder.create<fir::ArrayFetchOp>(
+              loc, eleTy, arrLd, iters.iterVec(), arrLdTypeParams);
+        }
+        if (auto eleTy = fir::dyn_cast_ptrEleTy(box.getType()))
+          if (!eleTy.isa<fir::BoxType>()) {
+            // `box` is not boxed. Embox it now.
+            box = builder.createBox(loc, box);
```

Comment:
Shouldn't `box` somehow be made an actual extended value before being passed to createBox, what about the length parameters ? I think you may need to call `arraySectionElementToExtendedValue` on `box` in this case to ensure the length parameters are set properly.

Reply:
Agreed on the extended value. Again, I don't think the present infrastructure is exactly what's needed.